# **1.Data understanding and EDA**

In [12]:
# Import Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [13]:
data=pd.read_csv('Datafiniti_Womens_Shoes.csv')

In [14]:
data.head()

id             dateAdded           dateUpdated asins  \
0  AVpfEf_hLJeJML431ueH  2015-05-04T12:13:08Z  2018-01-29T04:38:43Z   NaN   
1  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
2  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
3  AVpjXyCc1cnluZ0-V-Gj  2017-01-27T01:25:56Z  2018-01-04T11:52:35Z   NaN   
4  AVphGKLPilAPnD_x1Nrm  2017-01-27T01:25:56Z  2018-01-18T03:55:18Z   NaN   

         brand                                         categories  \
0  Naturalizer  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
1     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
2     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
3     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
4     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   

  primaryCategories                          colors                 dimension  \
0             Shoes  Silver,Cream Watercolor Floral                       NaN   
1             Shoes                            Grey                       NaN   
2             Shoes                            Grey                       NaN   
3       Shoes,Shoes                           Black  6.0 in x 6.0 in x 1.0 in   
4             Shoes                            Grey  6.0 in x 6.0 in x 1.0 in   

            ean  ...   prices.merchant prices.offer prices.returnPolicy  \
0           NaN  ...     Overstock.com          NaN                 NaN   
1  3.397705e+10  ...       Walmart.com          NaN                 NaN   
2  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
3  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
4  3.397705e+10  ...       Walmart.com          NaN                 NaN   

  prices.shipping prices.size  \
0             NaN           S   
1        Standard           6   
2           Value           6   
3           Value           6   
4       Expedited           6   

                                   prices.sourceURLs  \
0  https://www.overstock.com/Clothing-Shoes/Women...   
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   

                                               sizes  \
0  6W,9W,7.5W,12W,8.5M,9N,9M,9.5M,10.5M,10W,8.5W,...   
1                                         10,7,6,9,8   
2                                         10,7,6,9,8   
3                                         10,7,6,9,8   
4                                         10,7,6,9,8   

                                          sourceURLs           upc weight  
0  https://www.walmart.com/ip/Naturalizer-Danya-W...  017136472311    NaN  
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045903    NaN  
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045958    NaN  

[5 rows x 34 columns]

In [15]:
data.columns = data.columns.str.replace('.','_')

<ipython-input-15-7a0f35112492>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  data.columns = data.columns.str.replace('.','_')


In [16]:
data.head()

id             dateAdded           dateUpdated asins  \
0  AVpfEf_hLJeJML431ueH  2015-05-04T12:13:08Z  2018-01-29T04:38:43Z   NaN   
1  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
2  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
3  AVpjXyCc1cnluZ0-V-Gj  2017-01-27T01:25:56Z  2018-01-04T11:52:35Z   NaN   
4  AVphGKLPilAPnD_x1Nrm  2017-01-27T01:25:56Z  2018-01-18T03:55:18Z   NaN   

         brand                                         categories  \
0  Naturalizer  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
1     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
2     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
3     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
4     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   

  primaryCategories                          colors                 dimension  \
0             Shoes  Silver,Cream Watercolor Floral                       NaN   
1             Shoes                            Grey                       NaN   
2             Shoes                            Grey                       NaN   
3       Shoes,Shoes                           Black  6.0 in x 6.0 in x 1.0 in   
4             Shoes                            Grey  6.0 in x 6.0 in x 1.0 in   

            ean  ...   prices_merchant prices_offer prices_returnPolicy  \
0           NaN  ...     Overstock.com          NaN                 NaN   
1  3.397705e+10  ...       Walmart.com          NaN                 NaN   
2  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
3  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
4  3.397705e+10  ...       Walmart.com          NaN                 NaN   

  prices_shipping prices_size  \
0             NaN           S   
1        Standard           6   
2           Value           6   
3           Value           6   
4       Expedited           6   

                                   prices_sourceURLs  \
0  https://www.overstock.com/Clothing-Shoes/Women...   
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   

                                               sizes  \
0  6W,9W,7.5W,12W,8.5M,9N,9M,9.5M,10.5M,10W,8.5W,...   
1                                         10,7,6,9,8   
2                                         10,7,6,9,8   
3                                         10,7,6,9,8   
4                                         10,7,6,9,8   

                                          sourceURLs           upc weight  
0  https://www.walmart.com/ip/Naturalizer-Danya-W...  017136472311    NaN  
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045903    NaN  
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045958    NaN  

[5 rows x 34 columns]

In [17]:
data.shape

(10000, 34)

In [18]:
data.isna().sum()

id                         0
dateAdded                  0
dateUpdated                0
asins                   9997
brand                      0
categories                 0
primaryCategories          0
colors                  7369
dimension               9883
ean                     9329
imageURLs                  0
keys                       0
manufacturer            9473
manufacturerNumber      7518
name                       0
prices_amountMax           0
prices_amountMin           0
prices_availability     9566
prices_color               0
prices_condition        9562
prices_currency            0
prices_dateAdded         777
prices_dateSeen            0
prices_isSale              0
prices_merchant         9565
prices_offer            9879
prices_returnPolicy    10000
prices_shipping         9588
prices_size                0
prices_sourceURLs          0
sizes                      0
sourceURLs                 0
upc                      360
weight                  9701
dtype: int64

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 34 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   10000 non-null  object 
 1   dateAdded            10000 non-null  object 
 2   dateUpdated          10000 non-null  object 
 3   asins                3 non-null      object 
 4   brand                10000 non-null  object 
 5   categories           10000 non-null  object 
 6   primaryCategories    10000 non-null  object 
 7   colors               2631 non-null   object 
 8   dimension            117 non-null    object 
 9   ean                  671 non-null    float64
 10  imageURLs            10000 non-null  object 
 11  keys                 10000 non-null  object 
 12  manufacturer         527 non-null    object 
 13  manufacturerNumber   2482 non-null   object 
 14  name                 10000 non-null  object 
 15  prices_amountMax     10000 non-null  

In [20]:
data.describe(include='all')

id             dateAdded           dateUpdated  \
count                  10000                 10000                 10000   
unique                   653                   440                   495   
top     AV46uYe4glJLPUi8VW57  2017-11-10T19:33:19Z  2018-02-01T07:33:40Z   
freq                     185                   265                   235   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

             asins               brand                             categories  \
count            3               10000                                  10000   
unique           1                 129                                    207   
top     B016GSDNVM  journee collection  Womens,Shoes,Boots,Journee Collection   
freq             3                2182                                    835   
mean           NaN                 NaN                                    NaN   
std            NaN                 NaN                                    NaN   
min            NaN                 NaN                                    NaN   
25%            NaN                 NaN                                    NaN   
50%            NaN                 NaN                                    NaN   
75%            NaN                 NaN                                    NaN   
max            NaN                 NaN                                    NaN   

       primaryCategories                                             colors  \
count              10000                                               2631   
unique                 2                                                219   
top                Shoes  Blackchrome-Aurora Green-Hot Punch,Pure Platin...   
freq                9361                                                185   
mean                 NaN                                                NaN   
std                  NaN                                                NaN   
min                  NaN                                                NaN   
25%                  NaN                                                NaN   
50%                  NaN                                                NaN   
75%                  NaN                                                NaN   
max                  NaN                                                NaN   

                        dimension           ean  ...  prices_merchant  \
count                         117  6.710000e+02  ...              435   
unique                         19           NaN  ...               11   
top     8.0 in x 12.0 in x 4.0 in           NaN  ...  Backcountry.com   
freq                           82           NaN  ...              352   
mean                          NaN  7.714302e+11  ...              NaN   
std                           NaN  1.847119e+11  ...              NaN   
min                           NaN  3.397705e+10  ...              NaN   
25%                           NaN  7.276810e+11  ...              NaN   
50%                           NaN  8.701910e+11  ...              NaN   
75%                           NaN  8.860660e+11  ...              NaN   
max                           NaN  8.898850e+11  ...              NaN   

       prices_offer prices_returnPolicy  \
count           121                 0.0   
unique           11                 NaN   
top             35%                 NaN   
freq             23                 NaN   
mean            NaN                 NaN   
std             NaN                 NaN   
min          

1.The above shows numerical analysis of the given dataset

2.There are 10000 respondents

3.There are 129 unique brands and 207 Unique categories

4.The mean of Price_amount is 60.17 and Std is 18.8

In [21]:
data.columns

Index(['id', 'dateAdded', 'dateUpdated', 'asins', 'brand', 'categories',
       'primaryCategories', 'colors', 'dimension', 'ean', 'imageURLs', 'keys',
       'manufacturer', 'manufacturerNumber', 'name', 'prices_amountMax',
       'prices_amountMin', 'prices_availability', 'prices_color',
       'prices_condition', 'prices_currency', 'prices_dateAdded',
       'prices_dateSeen', 'prices_isSale', 'prices_merchant', 'prices_offer',
       'prices_returnPolicy', 'prices_shipping', 'prices_size',
       'prices_sourceURLs', 'sizes', 'sourceURLs', 'upc', 'weight'],
      dtype='object')

# Checking equality of similar named columns

We see some similar named columns let's check whether they are duplicate columns or not.

In [22]:
print('size',data['prices_size'].equals(data['sizes']))
print('color',data['prices_color'].equals(data['colors']))
print('sourceURL',data['prices_sourceURLs'].equals(data['sourceURLs']))

size False
color False
sourceURL False


They all seem to be different. We have two values of sizes(prices.size),colors(prices.color) and sourceURL(prices.sourceURLs) as we can see in the code above but for sake of simplicity we will not consider sizes,colors and sourceURL columns in our selected column list for our further processing. As we would want to keep these information with regards to the prices only.

# Choosing Columns
1.Since we don't have merchant information, columns containing unique identifier with respect to specific merchants, are not considered viz. asins (amazon id),keys (walmart id),manufacturerNumber (manufacturer id).

2.ean / upc columns are also not considered as for same product we can have multiple EANs (European Article Number) and UPCs (Universal Product Code) for smallest varaiations.

3.All Urls columns (imageURLs and sourceURLs)except prices_sourceURLs are not considered in selected columns as they are not relevant to our analysis.

In [23]:
selected_cols = ['id', 'dateUpdated', 'brand', 'categories','primaryCategories', 'dimension',
        'manufacturer', 'name','prices_amount' , 'prices_availability',
       'prices_color', 'prices_currency', 'prices_dateSeen', 'prices_isSale',
       'prices_shipping','prices_size','prices_sourceURLs']

Keeping a copy of the data from these selected columns into a new data frame called df, which we can continue to modify further without affecting the original data frame.

In [24]:
df = data[selected_cols].copy()

KeyError: "['prices_amount'] not in index"

In [ ]:
df.isna().sum()

From the above,we can see that 4 columns have more than 10% of null values.therefore we have to drop this 4 columns

# Extracting merchant information

prices_merchant was empty but we can still find the information from prices_sourceURLs.

In [ ]:
url = df['prices_sourceURLs'][0]
url

Now, parsing the url to get the merchant information

In [ ]:
print('step1:',url.split('www.'))
print('step2:',url.split('www.')[1])
print('step3:',url.split('www.')[1].split('/'))
print('step4:',url.split('www.')[1].split('/')[0])

Applying these above steps to all values of prices_sourceURLs with .apply() and seeing uniques merchants along with their counts using .value_counts()

In [ ]:
df['merchant'] = df['prices_sourceURLs'].apply(lambda x: x.split('www.')[1].split('/')[0])
df['merchant'].value_counts()

It is evident from above result that all prices are sourced from different merchants

# Merchant V/S Categories

In [ ]:
no_of_merchants=df.groupby('categories')['merchant'].count().sort_values(ascending=False)
no_of_merchants

From the above we can see that 'womens,Shoes,Boots,Journee Collection' category have highest number of merchants.And this is even more clear to observe from the below bar chart visualization which gives us the the top 10 categories based on number of merchants.

In [ ]:
plt.xticks(rotation=45)
plt.title("Number of merchant in 10 categories",fontsize=20)
plt.xlabel('categories')
plt.ylabel('No of merchants')
sns.barplot(no_of_merchants[:10],no_of_merchants[:10].index);

From above barchart we can clearly understand that Number of merchants are high in 'Womens,Shoes,Boots,Journee Collection' category.

# Prices_issale V/S Manufacturer

In [ ]:
df['prices_isSale'].value_counts()

In [ ]:
sale_df = df[df['prices_isSale'] != False]
print("sale_df's rows and columns:",sale_df.shape)
non_sale_df = df[df['prices_isSale'] == False]
print("non_sale_df's rows and columns:",non_sale_df.shape)

In [ ]:
#Before handling null values
top_manufacturers = non_sale_df.manufacturer.value_counts().head(10)
top_manufacturers

From the above result,we can understand that Dr.scholl's emerged as the top manufacturer by count of their product listing,being almost double th size of it's immediate competitor Brinley.Co.(Before handling missing values)

In [ ]:
#Before handling null values
top_manufacturers1 = sale_df.manufacturer.value_counts().head(10)
top_manufacturers1

In case of sale,'Brinley Co' category have the highest number of merchants.

In [ ]:
df.shape

# Brand V/S Prices_issale

In [ ]:
df['brand'].value_counts()

In [ ]:
print("No. of unique brands:",df.brand.nunique())

Let's look at the list of top 10 brands by count of their product listings

In [ ]:
top_brands = sale_df.brand.value_counts().head(10)
top_brands

In case of sale,'Brinley Co' brand is the top brand by count of their product listing.

In [ ]:
top_brands1 = non_sale_df.brand.value_counts().head(10)
top_brands1

In [ ]:
plt.figure(figsize=(12,6))
plt.xticks(rotation=45)
plt.title("Top 10 Women's Shoes Brands by their product listings",fontsize=20)
sns.barplot(top_brands.index, top_brands);

We can see from the visualization above that journee collection is almost double the size of it's immediate competitor. And this is even more clear to observe from the below pie chart visualization which gives us the actual market share by percentage between the top 5 brands.


In [ ]:
plt.figure(figsize=(12,6))
plt.title("Top 5 Brand share by percentage",)
plt.pie(top_brands[:5], labels=top_brands[:5].index, autopct='%1.1f%%', startangle=180);

The above pie chart shows how much percentage market is shared between these top 5 brands with a pie chart visualization.

# Brand V/S Avg Price

In [ ]:
av_price_brand_wise = df.groupby('brand')['prices_amount'].mean().sort_values(ascending=False)
av_price_brand_wise

In [ ]:
plt.xticks(rotation=45)
plt.title("Top 5 Women's Most Expensive Shoes Brands",fontsize=15)
sns.barplot(av_price_brand_wise[:5],av_price_brand_wise[:5].index);

from the above graph, we got red wing  turned out to be the most expensive brand based on average prices.

In [ ]:
av_price_brand_wise1 = df.groupby('brand')['prices_amount'].mean().sort_values(ascending=True)
av_price_brand_wise1

In [ ]:
plt.xticks(rotation=45)
plt.title("Top 5 Women's least Expensive Shoes Brands",fontsize=20)
sns.barplot(av_price_brand_wise1[:5],av_price_brand_wise1[:5].index);

From the above graph, we got Unbranded product turned out to be the least expensive brand on average prices.

In [ ]:
plt.figure(figsize=(12, 6))
plt.title("Price distribution")
plt.xlabel('Price bins')
plt.ylabel('Number brands falling in each bin')

plt.hist(df.prices_amount, bins=np.arange(15,315,5), color='orange');

In [ ]:
df.prices_amount.describe()

In [ ]:
plt.title("Price distribution",fontsize=20)
sns.distplot(df.prices_amount);

By looking at all of the above results, the answer is NO, distribution is no where close to normal. Infact, the Price distribution is rightly skewed suggesting to us that most of the brands prefer to price their products around the lower side of prices (viz. 15 to 50 USD) for better sales. Second bump is around 80 to 85 USD.

# Prices_issale V/S  prices_amount

In [ ]:
corrmatrix=df.corr()
plt.subplots(figsize=(20,8))
plt.title("Correlation between Price_issale and Prices_amount",fontsize=20)
sns.heatmap(corrmatrix,vmin=-.7,vmax=0.7,annot=True,linewidth=0.2)

The above shows correlation between Prices_amount and prices_issale.

# Brand V/S Color

In [ ]:
top_colors=df.prices_color.value_counts()
top_colors

In [ ]:
plt.figure(figsize=(12,6))
plt.xticks(rotation=45)
plt.xlabel('Colors')
plt.ylabel('Count')
plt.title("Top 5 Women's Shoes colors",fontsize=20)
sns.barplot(top_colors.index[:5], top_colors[:5]);

In [ ]:
# top 5 shoe colors
Black= df[df.prices_color =="Black"]
Grey= df[df.prices_color =="Gray"]
Taupe= df[df.prices_color =="Taupe"]
Navy= df[df.prices_color =="Navy"]
Brown=df[df.prices_color =="Brown"]

In [ ]:
#Black
Black.brand.value_counts()

In [ ]:
#Graph for Black color shoe collection - top 10 list
Black['brand'].value_counts()[:10].plot(kind='barh')
plt.title('Black color shoes data', fontsize = 20)
plt.xlabel('Count', fontsize = 10)
plt.ylabel('Brands', fontsize = 10)
plt.show()

The above shows black color shoes data.we can understand that the mostely used shoe color is black and also got that womens mostely used journee collection  black shoes

# Category V/S Avg Price

In [ ]:
av_price_category_wise = df.groupby(by='categories').mean()['prices_amount'].sort_values(ascending=False)
av_price_category_wise

In [ ]:
plt.xticks(rotation=45)
plt.title("Top 5 categories based on avg price",fontsize=20)
sns.barplot(av_price_category_wise[:5],av_price_category_wise[:5].index);

From the above barplot,we can understand that 'Women's Red Wing Shoes,Women's Clothing,Women's Shoes'emerged as the top category based on avg price.

In [ ]:
av_price_category_wise1 = df.groupby(by='categories').mean()['prices_amount'].sort_values(ascending=True)
av_price_category_wise1

In [ ]:
plt.xticks(rotation=45)
plt.title("Top 5 categories based on avg price",fontsize=20)
sns.barplot(av_price_category_wise1[:5],av_price_category_wise1[:5].index);

From the above barplot,we can understand that 'Clothing,Shoes,Womens Shoes,Womens Casual Shoes,Women's Shoes,Women's Casual Shoes,All Women's Shoes'emerged as the bottom category based on avg price.

# Stock availability

In [ ]:
df.prices_availability.value_counts()

After Assuming TRUE value  to be 'In stock',the availability of stock will be 406.Then the out of stock will be 28.